src/transport/types.ts

```diff
@@ -0,0 +1,73 @@
+import events from 'events'
+import Multiaddr from 'multiaddr'
+import Connection from '../connection/connection'
+import { Sink } from '../stream-muxer/types'
+
+export type DialOptions = {
+  signal?: AbortSignal
+}
+
+/**
+ * A libp2p transport is understood as something that offers a dial and listen interface to establish connections.
+ */
+export interface Transport <DialOptions extends { signal?: AbortSignal }> {
+  new (upgrader: Upgrader, ...others: any): Transport<DialOptions>; // eslint-disable-line
+  prototype: Transport <DialOptions>;
+  /**
+   * Dial a given multiaddr.
+   */
+  dial(ma: Multiaddr, options?: DialOptions): Promise<Connection>;
+  /**
+   * Create transport listeners.
+   */
+  createListener(options: any, handler: (Connection) => void): Listener;
```

Comment:
looks like `options` here should be also made into generic rather than any.

Reply:
I can change it to unknown, but really not much more. 
Taking a look at https://github.com/libp2p/js-libp2p-interfaces/tree/master/src/transport#create-a-listener there is no interface options for the listener. Each transport will have their own.

Reply:
But that also the whole point of generics to define a generic transport interface so that specific ones can specify concrete types that correspond to them just like it was in the case of dial:

```ts
interface Tansport<DialOptions extends { signal: AbortSignal }, ListenerOption> {
   dial(ma: Multiaddr, options?: DialOptions): Promise<Connection>;
   createListener(options: ListenerOptions, handler: (Connection) => void): Listener;
    filter(multiaddrs: Multiaddr[]): Multiaddr[];
}

class TCPTransport implements Transport<TCPDialOptions, TCPListenerOptions> {
   dial(ma: Multiaddr, options?: DialOptions): Promise<Connection> {
     // ...
   }
   createListener(options: TCPListenerOptions, handler: (connection: Connection) => void): Listener {
      // ...
   }
   // ...
}

type TCPDialOptions = {
  // ...
}

type TCPListenerOptions = //...
```